File: source/codes/leetcode/src/main/java/leetcode/L_198.go
package sample

func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	if len(nums) == 1 {
		return nums[0]
	}
	sums := make([]int, len(nums))
	sums[0] = nums[0]
	sums[1] = max(nums[0], nums[1])
	maxV := max(sums[0], sums[1])
	for i := 2; i < len(nums); i++ {
		sums[i] = max(sums[i-2]+nums[i], sums[i-1])
		maxV = max(sums[i], maxV)
	}

	return maxV
}

//func rob(nums []int) int {
//	s1 := robsum(nums, 0)
//	s2 := robsum(nums, 1)
//	if s2 > s1 {
//		return s2
//	}
//	return s1
//}
//
//func robsum(nums []int, cur int) int {
//	if cur >= len(nums) {
//		return 0
//	}
//	return nums[cur] + max(robsum(nums, cur+2), robsum(nums, cur+3))
//}

File: assets/codes/leetcode/src/main/java/top100liked/L_198.java
package top100liked;

public class L_198 {
    public int rob(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        if (nums.length == 2) {
            return Math.max(nums[0], nums[1]);
        }

        nums[1] = Math.max(nums[0], nums[1]);
        for (int i = 2; i < nums.length; i++) {
            nums[i] = Math.max(nums[i-1], nums[i-2]+nums[i]);
        }

        return nums[nums.length-1];

//
//        int[] dp = new int[nums.length];
//        dp[0] = nums[0];
//        dp[1] = Math.max(nums[0], nums[1]);
//
//        for (int i = 2; i < nums.length; i++) {
//            dp[i] = Math.max(dp[i - 2] + nums[i], dp[i - 1]);
//        }
//
//        return dp[nums.length - 1];
    }

    public static void main(String[] args) {
        L_198 l198 = new L_198();
        System.out.println(l198.rob(new int[]{1,2,3,1}));
        System.out.println(l198.rob(new int[]{2,7,9,3,1}));
    }
}
